### PR TITLE
Adds support for normalizing cuboid data before classifiying

### PR DESCRIPTION
**Pull request by matham**
## Description

This PR depends on https://github.com/brainglobe/cellfinder/pull/493. Only the last commit in this PR is unique to this PR.

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

It's fairly standard when distributing pre-trained models, in particular image networks, to indicate how the input should be normalized, if at all. E.g. `mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]` has been [used for ImageNet based models](https://docs.pytorch.org/tutorials/beginner/transfer_learning_tutorial.html#load-data) because that was the statistics of the ImageNet data. The reason is that the model will be more accurate if your data's statistics/input range matches the pre-trained dataset's statistics.

The pre-trained model distributed by brainglobe was trained on cells with implicit intensity statistics. So it would be nice if when using on other datasets we can normalize our own data to match that distribution. But it is not specified. Additionally, input data can vary wildly in intensity between channels and labeling types and so if you do transfer learning and want to provide a minimal set of new cells to train on, if you happen to fine tune on cells from images where the overall image is bright, trying to use it on images that are darker may result in worse classification.

When training on my own cells, I found that adding such normalization helped improve accuracy as you can see in the images below.

**What does this PR do?**

This PR add the option to normalize the training data, the classification data, and to save the image statistics into the yaml files when curating training data. Specifically,

1. During curation, it'll sample the z-stack of each channel and compute the mean/std and dump that into the "curated" yaml file. E.g. here's how the yaml file will look:
   ```py
   data:
   - bg_channel: 1
     cell_def: ''
     cube_dir: tests/data/integration/training/cells
     signal_channel: 0
     type: cell
     signal_mean: 241.31
     signal_std: 154.92
     bg_mean: 650.94
     bg_std: 217.90
   ```
   It adds the parameters `normalize_channels: bool = False` and `normalization_down_sampling: int = 32` to control this.
2. During training, we load these yaml files, and if selected, we normalize the cuboids by the mean/std of the dataset from which the cuboids originated. After training, the model expect input cuboids to have a mean of 1 and std of 1. This is controlled by the parameter `normalize_channels: bool = False`.
3. During classification, we similarly sample each channel in the z-stack and normalize the cuboids that want to classify by the mean/std of the corresponding channel. This is similarly controlled by the parameters `normalize_channels: bool = False` and `normalization_down_sampling: int = 32`.

By default all the normalization is turned OFF.

## References

https://github.com/brainglobe/cellfinder/pull/493.

## How has this PR been tested?

I tested this on my own curated data that was continued training from the pre-trained brainglobe models. As I was getting cellfinder classification to work on our c-fos and fostrap images I found the model wasn't performing good enough. So I added normalization which improved the model.

I re-did those experiments systematically on our training data just to illustrate the benefits. It seems that normalization is most helpful when you have less training data and when you're using a smaller learning rate. Both makes sense intuitively. For the former it's because if you don't have a lot of training data and they all have different statistics, the model has a harder time converging. For the same reason, a lower learning rate will likely get more easily stuck in local minima.

Here's the plotted train and test accuracy and loss for training on my data:

There are 4 plots, illustrating two different learning rates (`LR` = `1e-3` and `1e-4`), and two different values for the size of the training data (`f` = `10%` and `50%`, meaning the fraction of the training data used for testing - this is a quick way to reduce training data size by using `50%` for testing). `N=n` indicates the number of lines in the plot, which indicates the number of independent replications of the training. FYI, each replication uses a different train/test split, which explains the distribution of lines for a given condition.

But overall you can see how normalizing improves model performance, and this is only using training data from similar images (data augmentation was `25%` using `resnet50_tv` using a batch size of `32` and with `--continue-training`).

train_accuracy
test_accuracy
train_loss
test_loss

On the other hand, I tried to estimate the serial2p data's statistics from the provided cubes (already a problem because we want the statistics of the images they came from, not the extracted cubes) and saw little benefit when re-training the brainglobe model. I think this is because there's so much data in the training set that it really doesn't matter. Or the serial2p data are easier/cleaner to classify.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

There are new parameters. But, I think it may be worth to release a new pre-trained brainglobe model trained on normalized inputs.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: cellfinder/napari/curation.py [hunk 1]
```diff
@@ -632,21 +652,39 @@ def convert_layers_to_cells(self):
...
+        signal_stat = dataset_mean_std(
+            self.signal_layer.data, self.normalization_down_sampling
+        )
+        bg_stat = dataset_mean_std(
+            self.background_layer.data, self.normalization_down_sampling
+        )
+        return signal_stat, bg_stat
+
     def __save_yaml_file(self):
+        signal_stat, bg_stat = self._calculate_channel_stats()
```
Comment (alessandrofelder) [inline cellfinder/napari/curation.py:677]: Quick check that I have understood this correctly: we will from now on always save the per-channel stats after curation (and by default, we ignore them during training for backwards compatibility)? Presumably that won't affect the speed at which we save files much?

**Review by alessandrofelder — APPROVED** [review]
Thanks a lot for this @matham 

I have finally been able to review this 😅 

I confirm this is backwards compatible both in result and performance for our viral tracing `MS_cx_left` data. Noting results for posterity here:
```
Successfully installed cellfinder-1.10.1.dev1+ge127361a0
Running cellfinder from the main branch...
[...
Finished. Total time taken: 0:29:24.593917                            main.py:95

Successfully installed cellfinder-1.9.1.dev5+g59db9a492
Running cellfinder from the PR branch...
[...]
Finished. Total time taken: 0:29:23.058641                            main.py:95

Comparing the results...
Number of cells only found with cellfinder from main: 0
Number of cells only found with cellfinder from 588: 0
Number of matched cells: 70411
Matching proportion: 1.0
```
On the code changes side, all looks reasonable to me.
I just left one small question for something I wanted to double-check.

File: cellfinder/napari/curation.py (hunk 1 above)
Comment (matham) [reply cellfinder/napari/curation.py:677]: Correct! For my data of about 2000 giant planes, it take about a minute per channel to calculate the stats.

However...due to #548, the work happens in the main thread largely blocking the GUI. So while the stats are calculated for the couple of minutes it looks like the GUI is just frozen - you have to look at the terminal to see a progress bar. Once the stats are done, the progress bar in the GUI is updated with number of cells saved, even if the GUI stays frozen. Perhaps we need to add a message initially that stats are being calculated. Or maybe this can all be fixed if we move the work to a second thread?

**Comment by matham** [comment]
One consideration you may want to look at and let me know what you think. With this change, we simply sample every `n` planes. E.g. with `2000` planes down-sampled by `32`, we use `63` planes.

But for a smaller dataset, say `100` planes, the default `32` will only use `3` planes with 2 of them towards the start and end, which are probably dark planes. In this case, perhaps it would have been better if the parameter specified the number of planes we want to use in the sampling, say `10`, and from that we calculate the downs-sampling.

So e.g. if we defaulted it to `50`, then if the dataset was less than `50`, we used all planes. If it was `100`, we'd use every second plane. And for my dataset we'd use every 40th plane. Do you think this a better approach?

--------

Also, I was thinking, if possible the pre-trained model brainglobe provides should have a normalized option. This would mean the stats for the training data will have to be calculated from the original datasets and added to the training data. But this would make it possible for people to continue training with their data while using normalization, probably speeding up fine-tuning. Based on whether normalization is selected, cellfinder would download the correct weights. And ideally, long term, cellfinder would use normalization be default because there's no reason not to other than backwards compat. What do you think?

**Comment by alessandrofelder** [comment]
>So e.g. if we defaulted it to 50, then if the dataset was less than 50, we used all planes. If it was 100, we'd use every second plane. And for my dataset we'd use every 40th plane. Do you think this a better approach?

Yea, great point - I think this is a better approach - would you have capacity to fix in this PR?

This has made me realise: I guess this PR means that if we load a subset of a brain instead of the whole brain, but curate exactly the same in both cases, we will get different results due to the statistics of the image stack? (I think this is fine)

**Comment by alessandrofelder** [comment]
> Also, I was thinking, if possible the pre-trained model brainglobe provides should have a normalized option. This would mean the stats for the training data will have to be calculated from the original datasets and added to the training data. But this would make it possible for people to continue training with their data while using normalization, probably speeding up fine-tuning. Based on whether normalization is selected, cellfinder would download the correct weights. And ideally, long term, cellfinder would use normalization be default because there's no reason not to other than backwards compat. What do you think?

I agree, and have opened #617 to track (at least the first part) of your suggestion.

**Comment by matham** [comment]
>Yea, great point - I think this is a better approach - would you have capacity to fix in this PR?
Yes, will do soonish.

>This has made me realise: I guess this PR means that if we load a subset of a brain instead of the whole brain, but curate exactly the same in both cases, we will get different results due to the statistics of the image stack? (I think this is fine)

Yes! Assuming the stats of the subset is different enough and more importantly, the model wasn't trained with enough robustness/data to account for different data, you would get different results.

**Comment by matham** [comment]
I rebased and changed sampling factor to number of planes, as discussed!

File: cellfinder/napari/curation.py (hunk 1 above)
Comment (alessandrofelder) [reply cellfinder/napari/curation.py:677]: Opened #626 to take this discussion further. Maybe a progress bar in the GUI is sufficient here.

**Review by alessandrofelder — APPROVED** [review]
Awesome, thanks a lot @matham
